src/Zoom.vala

```diff
@@ -1,183 +1,180 @@
-                    wins.set_easing_duration (MOUSE_POLL_TIME);
-                    wins.pivot_point = new_pivot;
-                    wins.restore_easing_state ();
-                    return true;
-                });
+        gesture_tracker.connect_handlers (null, (owned) on_animation_update, null);
+    }
+
+    void zoom (float delta, bool play_sound, bool animate) {
+        // Nothing to do if zooming out of our bounds is requested
+        if ((current_zoom <= MIN_ZOOM && delta < 0) || (current_zoom >= MAX_ZOOM && delta >= 0)) {
+            if (play_sound) {
+                Gdk.beep ();
             }
+            return;
+        }
+
+        var wins = wm.ui_group;
 
-            current_zoom += delta;
-            var animation_duration = animate ? ANIMATION_DURATION : 0;
+        // Add timer to poll current mouse position to reposition window-group
+        // to show requested zoomed area
+        if (mouse_poll_timer == 0) {
+            float mx, my;
+            var client_pointer = Gdk.Display.get_default ().get_default_seat ().get_pointer ();
+            client_pointer.get_position (null, out mx, out my);
+            wins.set_pivot_point (mx / wins.width, my / wins.height);
 
-            if (current_zoom <= MIN_ZOOM) {
-                current_zoom = MIN_ZOOM;
+            mouse_poll_timer = Timeout.add (MOUSE_POLL_TIME, () => {
+                client_pointer.get_position (null, out mx, out my);
+                var new_pivot = Graphene.Point ();
 
-                if (mouse_poll_timer > 0)
-                    Source.remove (mouse_poll_timer);
-                mouse_poll_timer = 0;
+                new_pivot.init (mx / wins.width, my / wins.height);
+                if (wins.pivot_point.equal (new_pivot)) {
+                    return true;
+                }
 
                 wins.save_easing_state ();
-                wins.set_easing_mode (Clutter.AnimationMode.EASE_OUT_CUBIC);
-                wins.set_easing_duration (animation_duration);
-                wins.set_scale (MIN_ZOOM, MIN_ZOOM);
+                wins.set_easing_mode (Clutter.AnimationMode.LINEAR);
+                wins.set_easing_duration (MOUSE_POLL_TIME);
+                wins.pivot_point = new_pivot;
                 wins.restore_easing_state ();
+                return true;
+            });
+        }
 
-                if (animate) {
-                    wins_handler_id = wins.transitions_completed.connect (() => {
-                        wins.disconnect (wins_handler_id);
-                        wins.set_pivot_point (0.0f, 0.0f);
-                    });
-                } else {
-                    wins.set_pivot_point (0.0f, 0.0f);
-                }
+        current_zoom += delta;
+        var animation_duration = animate ? ANIMATION_DURATION : 0;
 
-                return;
+        if (wins_handler_id > 0) {
+            wins.disconnect (wins_handler_id);
+        }
+        wins_handler_id = 0;
```

Comment:
Shouldn't this come inside the if clause?

Reply:
Well, such checks here were done this way, so I followed it. I can fix this if you want.